crates/proto/src/lib.rs

```diff
@@ -88,3 +88,84 @@ pub use crate::xfer::retry_dns_handle::RetryDnsHandle;
 #[cfg(feature = "backtrace")]
 pub use error::{ENABLE_BACKTRACE, ExtBacktrace};
 pub use error::{ForwardData, ForwardNSData, ProtoError, ProtoErrorKind};
+
+#[cfg(not(feature = "no-std-rand"))]
+pub(crate) use rand::random;
+
+#[cfg(feature = "no-std-rand")]
+pub(crate) use no_std_rand::random;
+#[cfg(feature = "no-std-rand")]
+pub use no_std_rand::seed;
+
+/// A simple shim that allows us to use a [`random`] in `no_std` environments.
+#[cfg(feature = "no-std-rand")]
+mod no_std_rand {
+    use core::cell::RefCell;
+
+    use critical_section::Mutex;
+    use rand::distr::{Distribution, StandardUniform};
+    use rand::{Rng, SeedableRng, rngs::StdRng};
+
+    /// Generates a random value on `no_std`.
+    ///
+    /// # Panics
+    /// This function will panic if the rng has not been seeded.
+    /// The rng needs to be seeded using [`crate::seed`] before it can be used!
+    #[cfg(not(feature = "std"))]
+    pub(crate) fn random<T>() -> T
+    where
+        StandardUniform: Distribution<T>,
+    {
+        critical_section::with(|cs| {
+            RNG.borrow_ref_mut(cs)
+                .as_mut()
+                .expect("The no_std rng was not seeded using `hickory_proto::seed()`")
+                .random()
+        })
+    }
+
+    /// Generates a random value, shared behind a mutex.
+    /// This shared the API with its no_std counterpart.
+    ///
+    /// The no_std rng needs to be seeded using [`crate::seed`] before it can be used.
+    /// For testing purposes, if the std feature is enabled, the rng is seeded from the system,
+    /// but leaving it non-seeded would panic on no_std.
+    #[cfg(feature = "std")]
+    pub(crate) fn random<T>() -> T
+    where
+        StandardUniform: Distribution<T>,
+    {
+        critical_section::with(|cs| {
+            let mut rng_borrow = RNG.borrow_ref_mut(cs);
+
+            match rng_borrow.as_mut() {
+                Some(rng) => rng.random(),
+                None => {
+                    let mut rng = StdRng::from_os_rng();
```

Comment:
I've added a fallback that auto-seeds the rng in `std` to make tests pass (and also, why not)

Reply:
I'd prefer not to do this. I think we should guard the `no_std_rand` module existence on `not(feature = "std")` in addition to `feature = "no-std-rand"`. As for getting the tests passing, suggest doing `#[cfg(test)] seed(0x1337);`.

Reply:
Not every single one, but at least two tests were failing because of the missing seed.
I can add the seed to those, in case it's no-std (?)
But you see the issue

Reply:
I meant doing the `#[cfg(test)] seed()` inside the critical section closure, similar to what you did here.

Reply:
Ah, wait, you mean whenever there's std, we just do std?
gotcha

Reply:
Ah I might be confused again :)
I now just always use the std rand on std.
Really, there's never a reason to use anything else on std.
Please take a look